utils/src/poseidon/poseidon_hash.rs

```diff
@@ -81,20 +81,32 @@ impl<F: PrimeField> Poseidon<F> {
     }
 
     pub fn mix(&self, state: &[F], m: &[Vec<F>]) -> Vec<F> {
-        let mut new_state: Vec<F> = Vec::new();
+        let mut new_state: Vec<F> = vec![F::ZERO; state.len()];
         for i in 0..state.len() {
-            new_state.push(F::zero());
             for (j, state_item) in state.iter().enumerate() {
                 let mut mij = m[i][j];
                 mij *= state_item;
                 new_state[i] += mij;
             }
         }
-        new_state.clone()
+        new_state
+    }
+
+    pub fn mix_2(&self, state: &[F], m: &[Vec<F>], state_2: &mut [F]) {
+        // let mut new_state: Vec<F> = vec![F::ZERO; state.len()];
+        for i in 0..state.len() {
+            state_2[i] = F::ZERO;
```

Comment:
you set state_2 to zero on line 123, do we need to repeat it here?

Reply:
Yes it is required as in original fn mix(...), the new_state vector is initialized with F::ZERO